source/elements/oneMKL/source/domains/spblas/operations/spmv.rst

```diff
@@ -0,0 +1,378 @@
+     - Deterministic algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_CSR_ALG2``
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_stream``
+   * - ``csr_alg3``
+     - LRB variant of the algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: N/A
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_lrb``
+
+.. _onemkl_sparse_spmv:
+
+spmv
+----
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       void spmv_buffer_size(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
+       );
+
+       sycl::event spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           void*                                      workspace,
+           const std::vector<sycl::event>             &dependencies = {},
+       );
+
+       sycl::event spmv(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
```

Comment:
This may get a bit troublesome for the `sycl::buffer` backend. `alpha`/`beta` would need to be USM pointers even for the `sycl::buffer` APIs, and we don't like mixing `sycl::buffer` and USM.

For `sycl::buffer`s, having to track dependencies is normally not needed. If face a situation where `alpha`/`beta` must be discovered on the device along the way, then users would normally just use `sycl::buffer`s for `alpha`/`beta`. If they are forced to use USM pointers (because of the proposed `spmv` API for instance), they'll have to be careful about tracking dependencies, passing them on to `spmv`, and then the backend libs must ensure they insert those dependencies in the `spmv` kernel.

My initial opinion on this is to disallow device `alpha`/`beta` being pointers at all even though I know there are use-cases where `alpha`/`beta` aren't initially known. If we make it a `pointer`, then we should also explore the possibility of separating out a `buffer` API with `sycl::buffer<DATA_TYPE, 1> &alpha` argument.

@andrewtbarker, you worked on this for the BLAS domain, but I think y'all enabled it only for USM and not for buffers. Your thoughts on `alpha`/`beta` arguments here would be welcome.

Reply:
I think we should have alpha/beta be scalars for sycl::buffer APIs but allow them to be pointers or scalars for USM APIs

Reply:
I remember that there was some plan to allow scalars or pointers for alpha and beta parameters. I don't know what the status of that is. I think we should re-use that for the sparse domain once it's ready.

In the mean time the simplest solution is for alpha and beta parameters to either always be a scalar or always be a pointer. I am suggesting to use pointers for now as this can be a pointer on the host or on the device and it aligns better with cuSPARSE and rocSPARSE.

In any case I don't think it is worth making the API more complex for buffers.

Reply:
When we discussed this issue, I didn't realize that using `value_or_pointer` wouldn't work with the current API. We would need the types to be templated. I don't think it could be done in this version, see https://github.com/oneapi-src/oneAPI-spec/pull/522#discussion_r1548661139

Reply:
right, we will have trouble with value_or_pointer in these APIs since there is no distinguishing between USM and buffer APIs anymore (vectors are wrapped now) ...  we will need to think about what to do in the future about it ...  maybe overloads on interfaces ?  but won't be resolved in this PR change.  We can mark this is resolved for now.

Reply:
What is the decision here for now, though? Are we going to keep `void *alpha` or switch to `const FpType alpha`? If you keep `void *`, please reconsider and describe in the spec what the expected data types will need to be separately for `sycl::buffer` and `USM` cases, because the current description of `alpha`,
```
   alpha
      Host or USM pointer representing :math:`\alpha`. The USM allocation can be
      on the host or device. Must be of the same type than the handles' data
      type.
```
 will not work well for `sycl::buffer` because of the complication with dependencies that I described in my first review comment in this thread.

Reply:
I think we should keep `const void*` . I have clarified that the container types must be consistent in https://github.com/uxlfoundation/oneAPI-spec/pull/522/commits/1060c7a104d8390b2335ce7b82a1a1f3cdd035b5

Reply:
does that mean that we should have an argument similar to "cudaDataType              computeType"  that defines the data type that the operation is executed in ? See https://docs.nvidia.com/cuda/cusparse/index.html#cusparsespmv

Reply:
We will need an argument like this but it can be internal only. The user provides the type of the handles when creating the handle as a template argument. That type is stored internally as an enum and the wrapper can check that the types are consistent. For the `alpha` and `beta` arguments we can't do any type checking but we will know which type they need to be casted to thanks to the handles.
I have this type in my prototype for the CUDA backend [here](https://github.com/Rbiessy/oneMKL/pull/1/files#diff-b16bea1755ae34c4d99906fee1f9388609ebe53966f0a6d5451a21ef48f0aad1R35).